[GeoMechanicsApplication] Implement well constitutive behaviour for line thermal element

mnabideltares opened the pull request:
**📝 Description**
To enhance our computational model for simulating high capacity heat storage systems, we need to implement a thermal "wel/filter" element on top of the existing thermal line element. The thermal filter element will enable us to model the behavior of heat storage under the ground, accounting for their heat exchange processes. This addition is crucial for achieving realistic simulations of heat storage systems and optimizing their performance.

**🆕 Changelog**
- Added constutative model for the filter element
- Added test cases

mnabideltares committed:
Merge branch 'master' into geo/12322-implement-well-element

mnabideltares committed:
Added constutative behaviour

mnabideltares committed:
Merge branch 'master' into geo/12322-implement-well-element

mnabideltares committed:
Added test cases for the filter element

mnabideltares committed:
A fix in test case

mnabideltares committed:
README.md is added for the test cases

mnabideltares committed:
Fix in readme

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,51 @@
+SizeType GeoThermalFilterLaw::WorkingSpaceDimension() { return mNumberOfDimensions; }
+
+Matrix GeoThermalFilterLaw::CalculateThermalFilterMatrix(const Properties&  rProp,
+                                                                 const ProcessInfo& rProcessInfo) const
```
It seems clang-formatting does not work for you properly.

mnabideltares replied:
@markelov208  Thank you for the fast review and the interesting comment.
I don't know the reason why they call it filter in the field of geothermal, but this name seems to be common for thermal wells. However, if you have any suggestion for a better name, please let me know.

Indeed, my clang formatting sometimes doesn't function correctly. Maybe we can discuss/fix it this week.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,51 @@
+    Matrix result = ZeroMatrix(mNumberOfDimensions, mNumberOfDimensions);
+
+    const auto x = static_cast<int>(indexThermalFlux::X);
+    result(x, x) = rProp[THERMAL_CONDUCTIVITY_WATER];
```
As far as I understand mNumberOfDimensions is always equal to 1 and X is equal to 0. Then there is no need to use ZeroMatrix. From a general view-point, this part looks over complicated, for example, no need in mNumberOfDimensions. Please could you explain the reasons? Thank you.

markelov208 commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element/README.md:
```diff
@@ -0,0 +1,23 @@
+**Source files:** [Thermal filter element with fixed temperature](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element)
+
+## Case Specification
+In this thermal test case, a $2 \mathrm{[m]} \times 1 \mathrm{[m]}$ soil block is considered, with everywhere set to 0 $\mathrm{[^\circ C]}$. A well/filter passes vertically through middle of the domain. The temperature along the filter is kept constant at 0 $\mathrm{[^\circ C]}$. A sudden jump at the right and left boundary temperatures are given for the soil. These are both 25 $\mathrm{[^\circ  ...
```
"various configurations, including **2D3N** and **2D3N** soil elements", it seems one of them should be 2D2N, no? There is also an example with 2D6N below. 

"The temperature distribution is then evaluated with **its own result.**". Below it is written that the solution is compared with an analytical solution.

markelov208 commented on applications/GeoMechanicsApplication/tests/test_transient_thermal.py:
```diff
@@ -263,6 +263,14 @@ def test_thermal_point_flux_3D2N(self):
+    
+    def test_thermal_filter_element_2D2N(self):
+        temperature = self.simulate_thermal_case('test_thermal_filter_element/test_thermal_filter_element_2D2N')
+        self.assertAlmostEqual(7.679491925, temperature[22])
```
Please introduce a variable with clear name, for example, etalon_temperature_at_node80=7.67....

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,61 @@
+namespace Kratos
+{
+
+class RetentionLaw;
```
I do not see a reference on RetentionLaw here. Is it possible to remove this line?

WPK4FEM replied:
Is a RetentionLaw used?

markelov208 commented on applications/GeoMechanicsApplication/geo_mechanics_application.h:
```diff
@@ -601,6 +602,9 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) KratosGeoMechanicsApplication : publ
     const GeoThermalDispersionLaw mGeoThermalDispersion3DLaw{ConstitutiveLaw::SizeType(3)};
 
+    const GeoThermalFilterLaw mGeoThermalFilter2DLaw{ConstitutiveLaw::SizeType(1)};
+    const GeoThermalFilterLaw mGeoThermalFilter3DLaw{ConstitutiveLaw::SizeType(1)};
```
It is not clear for me to use **ConstitutiveLaw**::SizeType. Can it be just SizeType?

markelov208 reviewed (commented):
Hi Mohamed, thank you for the new feature. Sorry I have a few comments and a general question why is it called 'Thermal**Filter**'?

mnabideltares committed:
modifications based on review

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -31,18 +31,14 @@ ConstitutiveLaw::Pointer GeoThermalFilterLaw::Clone() const
-    result(x, x) = rProp[THERMAL_CONDUCTIVITY_WATER];
-
+    Matrix result = ZeroMatrix(1, 1);
+    result(0, 0) = rProp[THERMAL_CONDUCTIVITY_WATER];
```
There is no need to call ZeroMatrix because 0 will be over-written with rProp.... You may 
Matrix result(1,1); then 
result(0, 0) = rProp[THERMAL_CONDUCTIVITY_WATER];
or one line
auto result = ScalarMatrix(1,1,rProp[THERMAL_CONDUCTIVITY_WATER]);

If you would like to keep mNumberOfDimensions, then I will use it here and above instead of 1.

markelov208 reviewed (commented):
Hi Mohamed, concerning the Filter, let's keep it until @WPK4FEM  has a better suggestion.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,47 @@
+    : mNumberOfDimensions{NumberOfDimensions}
+{
+    KRATOS_ERROR_IF(mNumberOfDimensions != 1)
+        << "Got invalid number of dimensions: " << mNumberOfDimensions << std::endl;
```
If it always needs to be 1 exactly, the message should mention that.
It also implies that it is not necessary to have the number of dimensions as member variable, as it is not variable.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,47 @@
+    return Kratos::make_shared<GeoThermalFilterLaw>(*this);
+}
+
+SizeType GeoThermalFilterLaw::WorkingSpaceDimension() { return 1; }
```
When the WorkingSpaceDimension would not be constant, I expect that mNumberOfDimensions is returned here.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,61 @@
+class RetentionLaw;
+
+/**
+ * @class GeoThermalDispersionLaw
```
Copied comment? in this and the next lines.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/documentation_data/test_thermal_filter_element.svg:
```diff
@@ -0,0 +1 @@
+<svg width="1280" height="720" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" overflow="hidden"><defs><image width="1294" height="660" xlink:href="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEAYABgAAD/2wBDAAMCAgMCAgMDAwMEAwMEBQgFBQQEBQoHBwYIDAoMDAsKCwsNDhIQDQ4RDgsLEBYQERMUFRUVDA8XGBYUGBIUFRT/2wBDAQMEBAUEBQkFBQkUDQsNFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFB ...
```
The annotation Filter 0 degrees Celsius together with the result picture seem to imply the whole line is a 0. That would be a Dirichlet boundary condition, so probably you mean 0 at 1 point.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element/Common/MaterialParameters.json:
```diff
@@ -0,0 +1,48 @@
+        "SATURATED_SATURATION"             : 1.0,
+        "SPECIFIC_HEAT_CAPACITY_SOLID"     : 400.0,
+        "SPECIFIC_HEAT_CAPACITY_WATER"     : 3795.0,
+        "THERMAL_CONDUCTIVITY_SOLID_XX"    : 2.65,
```
Is the code always looking at XX for the solid contribution, XX seems to refer to the global X direction, while mostly you show pictures of vertical "filters".

mnabideltares replied:
It will be done in a new PR to avoid "if-else" statements in the thermal element. A new issue is added to the product backlog:

[GeoMechanicsApplication] Move the check_properties of the material to the constutative model for thermal element

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element/README.md:
```diff
@@ -0,0 +1,22 @@
+**Source files:** [Thermal filter element with fixed temperature](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element)
+
+## Case Specification
+In this thermal test case, a $2 \mathrm{[m]} \times 1 \mathrm{[m]}$ soil block is considered, with everywhere set to 0 $\mathrm{[^\circ C]}$. A well/filter passes vertically through middle of the domain. The temperature along the filter is kept constant at 0 $\mathrm{[^\circ C]}$. A sudden jump at the right and left boundary temperatures are given for the soil. These are both 25 $\mathrm{[^\circ  ...
```
Also here you write about a Dirichlet condition. What is then the purpose of the line elements

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element/test_thermal_filter_element_2D2N/test_thermal_filter_element_2D2N.mdpa:
```diff
@@ -0,0 +1,629 @@
+Begin ModelPartData
+//  VARIABLE_NAME value
+End ModelPartData
+
```
What is the purpose of these first 3 lines?

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element/test_thermal_filter_element_2D3N/test_thermal_filter_element_2D3N.mdpa:
```diff
@@ -0,0 +1,1162 @@
+Begin ModelPartData
+//  VARIABLE_NAME value
+End ModelPartData
+
```
Same question as for the 2D. What is the purpose of the first 3 lines.

WPK4FEM reviewed (changes requested):
Hi Mohammed,
I have some trouble understanding the test. A Dirichlet condition at the same place will dictate the solution, and as such prevent testing of the constitutive behaviour in the line thermal element.

Also the constitutive law only looks at the conductivity of water, If that is all that is used, much of the given input for the filter element can be removed.

Regards, Wijtze Pieter

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,47 @@
+
+GeoThermalFilterLaw::GeoThermalFilterLaw() : mNumberOfDimensions{1} {}
+
+GeoThermalFilterLaw::GeoThermalFilterLaw(std::size_t NumberOfDimensions)
```
It would be good to add unit tests for this new class, it seems quite doable since there is relatively little setup needed for testing the functions

mnabideltares committed:
Modifications based on review comments

mnabideltares committed:
Fix based on review

rfaasse commented on applications/GeoMechanicsApplication/tests/test_thermal_element/documentation_data/test_thermal_filter_element_2D3N_result.png:
Would it be a lot of work to export an svg for this one? It is okay like this, it's just a suggestion if it's simple to do

mnabideltares replied:
@rfaasse Thank you for the review.
I exported it from GiD. I couldn't find possibility in GiD to export as svg. Used Gimp, it seems it doesn't convert png to svg. Please let me know if you have a better solution/idea.

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,45 @@
+    return Kratos::make_shared<GeoThermalFilterLaw>(*this);
+}
+
+SizeType GeoThermalFilterLaw::WorkingSpaceDimension() { return mNumberOfDimensions; }
```
I'm starting to doubt now if we maybe should use SizeType after all, since we override this function from Constitutive Law

rfaasse reviewed (commented):
Clear PR and nice integration test cases with clear documentation. I think it would be nice to add unit tests for the new class as well. Since it seems not a lot of setup is necessary, this should be doable. Other than that, I only have a few very minor remarks,

mnabideltares committed:
fixes based on review comments

mnabideltares committed:
fixes based or review

mnabideltares committed:
Fix in unit-test for thermal filter constitutive law

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,51 @@
+    GeoThermalDispersionLaw geo_thermal_filter_law(dimension);
+
+    KRATOS_EXPECT_EQ(geo_thermal_filter_law.WorkingSpaceDimension(), dimension);
+}
```
We could add a simple test to check that the construction throws when the dimension is wrong by using `KRATOS_EXPECT_EXCEPTION_IS_THROWN`, see e.g.
https://github.com/KratosMultiphysics/Kratos/blob/1166a00ee1a1d91e2019706e5043e610d4c58807/applications/GeoMechanicsApplication/tests/cpp_tests/test_variables_utilities.cpp#L26-L31

I don't consider this blocking, but it might be a nice test to add.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,51 @@
+
+namespace Kratos::Testing {
+
+KRATOS_TEST_CASE_IN_SUITE(CalculateThermalFilterLawMatrix, KratosGeoMechanicsFastSuite)
```
Thanks for adding these unit tests!

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,51 @@
+    const Matrix thermal_filter_matrix = geo_thermal_filter_law.CalculateThermalFilterMatrix(*cond_prop, info);
+
+    Matrix expected_solution = ZeroMatrix(1, 1);
+    expected_solution(0, 0)  = 1000.0;
```
This might simplify a bit, but this is personal preference, so leaving this up to you.
```suggestion
    Matrix expected_solution = ScalarMatrix(1, 1, 1000.0);
```

rfaasse commented on applications/GeoMechanicsApplication/geo_mechanics_application.h:
```diff
@@ -601,6 +602,9 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) KratosGeoMechanicsApplication : publ
     const GeoThermalDispersionLaw mGeoThermalDispersion3DLaw{ConstitutiveLaw::SizeType(3)};
 
+    const GeoThermalFilterLaw mGeoThermalFilter2DLaw{std::size_t(1)};
+    const GeoThermalFilterLaw mGeoThermalFilter3DLaw{std::size_t(1)};
```
Should this be reverted to ConstitutiveLaw::SizeType as well?

rfaasse reviewed (dismissed):
No blocking comments from my side, thanks for making the improvements and adding the tests! I have a few suggestions you may or may not want to take up.

mnabideltares committed:
Addressed review comments

mnabideltares committed:
clang format

mnabideltares committed:
THERMAL_LAW is added, GeoThermalLaw base class is defined. Dispersion and Filter laws changed inheritance to GeoThermalLaw

Test cases are modified, as the constutative law for heat is removed, and replaced by thermal law

mnabideltares committed:
In the case of missing thermal_law, the dispersion-law is set as default

fixed code smell

mnabideltares committed:
fixed code smells

mnabideltares committed:
fixes to avoid WError

mnabideltares committed:
Fixes for WError's

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -269,9 +270,25 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientThermalElement : public Ele
         const ProcessInfo&                               rCurrentProcessInfo) const
     {
         const std::size_t number_of_dimensions = GetGeometry().LocalSpaceDimension();
+        unique_ptr<GeoThermalLaw> law;
```
Minor suggestion
```suggestion
        std::unique_ptr<GeoThermalLaw> law;
```

mnabideltares replied:
@rfaasse Thank you for your help and for reviewing. I have made all the changes you suggested. Then let's wait for the others opinion.

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_law.h:
```diff
@@ -0,0 +1,61 @@
+
+    virtual ~GeoThermalLaw() = default;
+
+    virtual GeoThermalLaw::Pointer Clone() const = 0;
```
I don't think we're still using the Clone and WorkingSpaceDimension functions anywhere. If that's indeed the case, I'd suggest to remove the functions

avdg81 replied:
Indeed, member `WorkingSpaceDimension` is only called by a unit test, but it is not being used elsewhere. Also, I have verified that when the `Clone` member is removed, all still builds. So apparently, it is not being called. So please, remove these two redundant members. Thank you.

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.h:
```diff
@@ -36,26 +36,24 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoThermalDispersionLaw : public Con
     {
-        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, ConstitutiveLaw)
+        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, GeoThermalLaw)
         rSerializer.load("NumberOfDimensions", mNumberOfDimensions);
```
I think we're now saving and loading the mNumberOfDimensions twice (once in the base and once in the derived classes). I think it would be best to not have a protected member mNumberOfDimensions, but just have one in both the derived classes instead. Protected variables are counted as code smells.

avdg81 replied:
Yes, I agree. Class `GeoThermalLaw` should only provide an abstract interface and not contain any data. The data should be kept by the concrete classes, such as `GeoThermalDispersionLaw`. And please, make the data `private` there, not `protected`.

rfaasse reviewed (commented):
Thank you for your hard work on this Mohamed! It's great to see we can read it from json in this way. 

I have a few minor comments left, but nothing blocking. However, since I was quite involved in some of the ideas, I will not approve the PR and prefer for someone else to do this in the end (since I'm of course a bit biased).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.h:
```diff
@@ -36,26 +36,24 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoThermalDispersionLaw : public Con
     explicit GeoThermalDispersionLaw(SizeType NumberOfDimensions);
 
-    ConstitutiveLaw::Pointer Clone() const override;
+    ~GeoThermalDispersionLaw() override = default;
```
Destructors should not be marked as `override` (they are the exception to the rule when `virtual` functions are overridden). In fact, since you don't have any special semantics that require you to explicitly define copy and move operations, I would leave out the explicit definition of a `default` destructor. In fact, by explicitly defining it, you suppress the default move operations. So please, apply the [rule of zero](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c20-if-you-can-avoid-defining-default-operations-do) whenever you can.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,57 @@
+
+    explicit GeoThermalFilterLaw(SizeType NumberOfDimensions);
+
+    ~GeoThermalFilterLaw() override = default;
```
This definition is not needed (see one of my previous comments).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_law.h:
```diff
@@ -0,0 +1,61 @@
+    virtual void load(Serializer& rSerializer)
+    {
+        rSerializer.load("NumberOfDimensions", mNumberOfDimensions);
+    }
```
These lines can be removed after moving`mNumberOfDimensions` to the concrete classes. Since this abstract base class only provides an interface (and thus it doesn't contain any data), no serialization is needed.

mnabideltares committed:
Addressed changes based on review

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.h:
```diff
@@ -36,26 +36,24 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoThermalDispersionLaw : public Con
 
 private:
-    SizeType mNumberOfDimensions;
+    std::size_t mNumberOfDimensions = 0;
```
In line with the behavior of the default constructor, I would change the default value to 2 instead of 0. In that case, you can remove the implementation of the default constructor from the `cpp` file and append `= default` to its definition. Here you are required to do that, since you also define a non-default constructor.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.h:
```diff
@@ -36,26 +36,24 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoThermalDispersionLaw : public Con
     void save(Serializer& rSerializer) const override
     {
-        KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, ConstitutiveLaw)
+        KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, GeoThermalLaw)
```
No longer needed if you follow my earlier suggestion. The abstract base class shouldn't have any data, and therefore doesn't need to be serialized.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.h:
```diff
@@ -36,26 +36,24 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoThermalDispersionLaw : public Con
     void load(Serializer& rSerializer) override
     {
-        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, ConstitutiveLaw)
+        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, GeoThermalLaw)
```
Also this line should no longer be needed.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,35 @@
+    KRATOS_ERROR_IF(mNumberOfDimensions != 1)
+        << "Got invalid number of dimensions. The dimension has to be 1, but got: " << mNumberOfDimensions
+        << std::endl;
+}
```
Since the only acceptable number of dimensions equals 1, I don't see the point of having this constructor. In fact, `mNumberOfDimensions` is never used by this class, so I would remove that data member. As a result, the custom default constructor has become redundant, too.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.cpp:
```diff
@@ -0,0 +1,35 @@
+namespace Kratos
+{
+
+GeoThermalFilterLaw::GeoThermalFilterLaw() : mNumberOfDimensions{1} {}
```
Probably, this one is no longer needed. See my other comment.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,57 @@
+
+    GeoThermalFilterLaw();
+
+    explicit GeoThermalFilterLaw(SizeType NumberOfDimensions);
```
These are both no longer needed, I reckon.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,57 @@
+    Matrix CalculateThermalDispersionMatrix(const Properties& rProp, const ProcessInfo&) const override;
+
+private:
+    std::size_t mNumberOfDimensions = 0;
```
This data member is not being used, so it can be removed.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,57 @@
+    {
+        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, GeoThermalLaw)
+        rSerializer.load("NumberOfDimensions", mNumberOfDimensions);
+    }
```
I would suggest to make these function bodies empty (since there won't be any data members anymore, and also the abstract base class doesn't support serialization either).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_law.h:
```diff
@@ -0,0 +1,48 @@
+
+    virtual void save(Serializer& rSerializer) const {}
+
+    virtual void load(Serializer& rSerializer) {}
```
I would make these pure virtual, since this abstract base class doesn't provide any serialization by itself. It just provides the interface, nothing else.
```suggestion
    virtual void save(Serializer& rSerializer) const = 0;

    virtual void load(Serializer& rSerializer) = 0;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_law.h:
```diff
@@ -0,0 +1,48 @@
+    /// Counted pointer of GeoThermalLaw
+    KRATOS_CLASS_POINTER_DEFINITION(GeoThermalLaw);
+
+    GeoThermalLaw() {}
```
I guess that this one was only needed due to the serialization. Now that serialization is no longer implemented by this class, I guess we no longer need a default constructor. But please, try and see whether my statement is correct.

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -269,9 +270,25 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientThermalElement : public Ele
+                law = std::make_unique<GeoThermalDispersionLaw>(number_of_dimensions);
+            } 
+            else if (rThermalLawName == "GeoThermalFilterLaw") {
+                law = std::make_unique<GeoThermalFilterLaw>(number_of_dimensions);
```
Since the thermal filter law doesn't use the number of dimensions (at present, it only supports one-dimensional spaces), we can do with a default constructor (as I previously suggested).

Also, I would prefer to factor out the code that creates the thermal law in a private member function, e.g. `CreateThermalLaw`. Making the thermal law then reduces to:
```
auto law = CreateThermalLaw();
```

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -378,6 +376,7 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_VARIABLE( DT_TEMPERATURE_COEFFICIENT )
     KRATOS_REGISTER_VARIABLE( DT_TEMPERATURE )
     KRATOS_REGISTER_VARIABLE( NORMAL_HEAT_FLUX )
+    KRATOS_REGISTER_VARIABLE( THERMAL_LAW)
```
Wouldn't `THERMAL_LAW_NAME` be more accurate?

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp:
```diff
@@ -101,12 +101,9 @@ KRATOS_TEST_CASE_IN_SUITE(CalculateThermalDispersionMatrix3D, KratosGeoMechanics
 }
 
-KRATOS_TEST_CASE_IN_SUITE(GetWorkingSpaceDimension_ReturnsCorrectValue, KratosGeoMechanicsFastSuite)
+KRATOS_TEST_CASE_IN_SUITE(TestDispersionLawThrowsWhenDimensionInvalid, KratosGeoMechanicsFastSuite)
```
This test becomes redundant when the number of dimensions can no longer be passed to the constructor.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,48 @@
+    auto cond_prop = r_model_part.CreateNewProperties(0);
+    cond_prop->SetValue(THERMAL_CONDUCTIVITY_WATER, 1000.0);
+
+    const SizeType      dimension = 1;
```
This variable becomes redundant, I reckon.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,48 @@
+    Model current_model;
+    auto& r_model_part = current_model.CreateModelPart("ModelPart");
+
+    auto cond_prop = r_model_part.CreateNewProperties(0);
```
In line with the Kratos Style Guide, we should rename this variable to `p_cond_prop`.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,48 @@
+
+    constexpr double tolerance{1.0e-6};
+
+    KRATOS_EXPECT_NEAR(thermal_filter_matrix(0, 0), expected_solution(0, 0), tolerance);
```
I would change this check to a matrix-based one (which also checks the dimensions):
```suggestion
    KRATOS_EXPECT_MATRIX_NEAR(thermal_filter_matrix, expected_solution, tolerance);
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_filter_law.cpp:
```diff
@@ -0,0 +1,48 @@
+{
+    KRATOS_EXPECT_EXCEPTION_IS_THROWN(GeoThermalFilterLaw law{2},
+        "Got invalid number of dimensions. The dimension has to be 1, but got: 2")
+}
```
This unit test should no longer be needed, if you remove the constructor that accepts a number of dimensions (which must always be equal to 1).

avdg81 commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_micro_climate_1/MaterialParameters.json:
```diff
@@ -25,7 +22,8 @@
         "SMIN_COEFFICIENT"                 : 0.000000e+00,
-        "SMAX_COEFFICIENT"                 : 4.800000e-04
+        "SMAX_COEFFICIENT"                 : 4.800000e-04,
+        "THERMAL_LAW"                      : "GeoThermalDispersionLaw"
```
Since this input parameter determines which other parameters are needed, I would put this as the very first one in the list.

avdg81 reviewed (changes requested):
This is a very nice extension to the work that you have done for geothermal analysis. Although the basis of your extension is good, I still have a few suggestions that would make the extension even simpler and cleaner. Hope you find my feedback useful.

mnabideltares committed:
Modifications based on review

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_filter_law.h:
```diff
@@ -0,0 +1,35 @@
+
+    void save(Serializer&) const override {}
+
+    void load(Serializer&) override {}
```
Please add comments to these function bodies which explain why they are empty. For instance, "Nothing to serialize, since there are no data members, and its base class is abstract". This will resolve two code smells found by SonarQube.

avdg81 commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element/README.md:
```diff
@@ -0,0 +1,17 @@
+**Source files:** [Thermal filter element with fixed temperature](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_filter_element)
+
+## Case Specification
+In this thermal test case, a $2 \mathrm{[m]} \times 1 \mathrm{[m]}$ soil block is considered, with everywhere set to 0 $\mathrm{[^\circ C]}$. A well/filter passes vertically through middle of the domain. Heat fluxes are set at the top and bottom of the filter, with values of 100 $\mathrm{[W/m^2]}$. Dirichlet boundary donditions of temperature are set at the left and right of the soil domain, both ...
```
Typo: "donditions" => "conditions"

WPK4FEM replied:
Like m and \circ C, days indicates a unit and should have [ ].

avdg81 commented on applications/GeoMechanicsApplication/tests/test_thermal_element/test_thermal_fixed_temperature/Common/MaterialParameters2D.json:
```diff
@@ -17,7 +15,7 @@
               "THERMAL_CONDUCTIVITY_SOLID_YY" :  1.50,
               "THERMAL_CONDUCTIVITY_SOLID_XY" :  0.0,
-              "THERMAL_CONDUCTIVITY_WATER"    :  0.48
+              "THERMAL_CONDUCTIVITY_WATER"    :  0.48              
```
This trailing spaces were probably inserted accidentally.

avdg81 reviewed (dismissed):
After your latest changes I'm very satisfied about this PR. Note that I have only a few minor remarks left. Please, also have a look at the code smells reported by SonarQube (there are 3 left, which are quite easy to fix). Thank you.

P.S.: You may want to wait with merging until @WPK4FEM has had another look.

WPK4FEM reviewed (dismissed):
Thank you for the tedious work, this looks good.

mnabideltares committed:
minor fixes based on review

mnabideltares committed:
minor fix for code smell

avdg81 reviewed (approved):
I believe this is good to go. Thanks for the work done Mohamed!
